Inverse Gaussian sampling

xshi19 opened the pull request:
I tried to use NP directly in sampling instead of converting lambda and mu from EP. Somehow it works:
```
lambda_ = 3.0
mu = 5.0
true_params = InverseGaussianNP(
    negative_lambda_over_two_mu_squared=jnp.array(-lambda_ / (2 * mu**2)),
    negative_lambda_over_two=jnp.array(-lambda_ / 2)
)

n_samples = 10000
samples_nat = true_params.sample(jr.split(key)[0], (n_samples,)) # new sample under NP
print(samples_nat.mean()) # 4.9460945

true_exp = true_params.to_exp()
samples_exp = true_exp.sample(jr.split(key)[0], (n_samples,))  # original sample under EP
print(samples_exp.mean()) # 9.204897
```

NeilGirdhar commented:
Oh, you found my bug!!  Thank you.  I forgot to split the key.

Can you also remove the line in `conftest.py` where the sampling test is blocked for `InverseGaussian`?

xshi19 commented:
> Oh, you found my bug!! Thank you. I forgot to split the key.
> 
> Can you also remove the line in `conftest.py` where the sampling test is blocked for `InverseGaussian`?

Yes I removed the lines: https://github.com/NeilGirdhar/efax/pull/39/commits/90f8d19abad25e3dbb93d4273c431415e086a205

NeilGirdhar commented:
Looks great, can you rebase off main and squash?  I'll merge right after!

xshi19 commented:
> Looks great, can you squash (`git rebase --interactive @~20`) and rebase off main (`git rebase --onto main @~1`)? I'll merge right after!

The rebase process should be done, can you take a look?

NeilGirdhar commented:
> The rebase process should be done, can you take a look?

Did you forget to force-push your changes after squashing and rebasing?  This is what your repo is showing:

```
❯ git log --oneline --decorate --color --graph --all
*   9bb6d87 (HEAD -> invgauss, origin/invgauss) Merge branch 'invgauss' of github.com:xshi19/efax into invgauss
|\  
| *   030f806 Merge branch 'invgauss' of github.com:xshi19/efax into invgauss
| |\  
| | * 9aeb73e resolve conflicts
| | * 31d7e5a resolve conflicts
| | * 90f8d19 Remove commented-out code related to Inverse Gaussian in pytest_generate_tests
| * | 547f4b1 resolve conflicts
| |/  
| * 1fbb08b change sample using NP directly
| * 34321b1 (origin/main, origin/HEAD, main) Bump
| * 9a77b3e Update examples
| * ec6f020 Eliminate Jax initialization on import
| * ed70612 Bump
* | 55a0b73 Update lock
| | * c8d186a (origin/diagonalize) Add a mixin that automatically diagonalizes a distribution
| |/  
| | * 1557590 (origin/new_infra) Switch to scipy's new distribution infrastructure
| |/  
| | * b54fc78 (origin/wishart) Add the Wishart distribution
| |/  
| * 0662823 Correct docstrings
| * 0797528 Split parameter package
| * e57bfe1 Correct issue with scalar support bound shapes
| * 23ad165 Skip sampling test for inverse Gaussian
| * 3fe090b Factor out triangular number functions
| * 379af3f Polish parameter_supports mecahnism
| * ff71ec7 Split structure classes into 3 files
| * e2f68db (tag: v1.21.0) Bump
| * e876d05 Add the inverse Gaussian distribution
| * fa6c27a Update RealField to work with arrays
| * d3ae652 Correct typo in test
| | * 86697f5 (origin/segfault) Change illustrating segmentation fault
| |/  
| | * d6b29fa (origin/use_gamma_hessian) Rename classes and move test to new file
| |/  
| * 7553c4e Use conditional expression with shape
| * 7c35664 Update lock
|/  
* d605f45 Use reciprocal
```
It should look like this:
```
* 1fbb08b (HEAD -> invgauss, origin/invgauss) change sample using NP directly
* 34321b1 (origin/main, origin/HEAD, main) Bump
* 9a77b3e Update examples
* ec6f020 Eliminate Jax initialization on import
* ed70612 Bump
| 55a0b73 Update lock
| * c8d186a (origin/diagonalize) Add a mixin that automatically diagonalizes a distribution
|/  
| * 1557590 (origin/new_infra) Switch to scipy's new distribution infrastructure
|/  
| * b54fc78 (origin/wishart) Add the Wishart distribution
|/  
* 0662823 Correct docstrings
* 0797528 Split parameter package
* e57bfe1 Correct issue with scalar support bound shapes
* 23ad165 Skip sampling test for inverse Gaussian
* 3fe090b Factor out triangular number functions
* 379af3f Polish parameter_supports mecahnism
* ff71ec7 Split structure classes into 3 files
* e2f68db (tag: v1.21.0) Bump
* e876d05 Add the inverse Gaussian distribution
* fa6c27a Update RealField to work with arrays
* d3ae652 Correct typo in test
| * 86697f5 (origin/segfault) Change illustrating segmentation fault
|/  
| * d6b29fa (origin/use_gamma_hessian) Rename classes and move test to new file
|/  
* 7553c4e Use conditional expression with shape
* 7c35664 Update lock
* d605f45 Use reciprocal
```
If you haven't already, could you try to squash and then rebase off main (commands above), and then force push: `git push --force-with-lease`?

If you `git log`, it should end with:
```
* 1fbb08b (HEAD -> invgauss, origin/invgauss) change sample using NP directly
* 34321b1 (origin/main, origin/HEAD, main) Bump
```

NeilGirdhar commented:
Okay, I was able to make the changes on your branch directly.  Thank you very much for this contribution, I really was not looking forward to debugging this!

If you're curious what I did to fix the branch, I ended up doing a hard reset to your first change (to clear away the confusing merges), I added the upstream pointer (`git remote add upstream ...`) then I rebased off `upstream/main`: `git rebase --onto upstream/main @~1`.  Then I redid the change to `conftest.py`.

xshi19 commented:
Thanks for helping me push the changes. I did not force push, but tried to
resolve the conflicts after squashing and rebasing. It was a bit confusing
for me.


On Tue, Apr 22, 2025 at 9:05 PM Neil Girdhar ***@***.***>
wrote:

> Okay, I was able to make the changes on your branch directly. Thank you
> very much for this contribution, I really was not looking forward to
> debugging this, so I greatly appreciate your contribution/help!
>
> If you're curious what I did to fix the branch, I ended up doing a hard
> reset to your first change (to clear away the confusing merges), I added
> the upstream pointer (git remote add upstream ...) then I rebased off
> upstream/main: git rebase --onto upstream/main @~1. Then I redid the
> change to conftest.py.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NeilGirdhar/efax/pull/39#issuecomment-2822812836>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADDLVP5FP4SC3CIFVDWFDPT223RNBAVCNFSM6AAAAAB3J25IYOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMRSHAYTEOBTGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> *NeilGirdhar* left a comment (NeilGirdhar/efax#39)
> <https://github.com/NeilGirdhar/efax/pull/39#issuecomment-2822812836>
>
> Okay, I was able to make the changes on your branch directly. Thank you
> very much for this contribution, I really was not looking forward to
> debugging this, so I greatly appreciate your contribution/help!
>
> If you're curious what I did to fix the branch, I ended up doing a hard
> reset to your first change (to clear away the confusing merges), I added
> the upstream pointer (git remote add upstream ...) then I rebased off
> upstream/main: git rebase --onto upstream/main @~1. Then I redid the
> change to conftest.py.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NeilGirdhar/efax/pull/39#issuecomment-2822812836>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADDLVP5FP4SC3CIFVDWFDPT223RNBAVCNFSM6AAAAAB3J25IYOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMRSHAYTEOBTGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

NeilGirdhar commented:
No worries, I used to find git really confusing.  In the future, I suggest going on #git on IRC for help from friendly people.
